util/util_unix.go

```diff
@@ -0,0 +1,31 @@
+// +build linux darwin
+
+package util
+
+import (
+	"os"
+	"sync"
+	"syscall"
+)
+
+type hardlinkDeviceAndInode struct {
+	device, inode uint64
+}
+
+type HardlinkChecker struct {
+	hardlinks sync.Map
+}
+
+func (h *HardlinkChecker) Check(fi os.FileInfo) string {
+	if stat, ok := fi.Sys().(*syscall.Stat_t); ok && stat.Nlink > 1 {
```

Comment:
Should you avoid directories?

Reply:
For correctness, yes.  We shouldn't call this except for regular files, but still.